New "atmanlfv3inc" Rocoto job

DavidNew-NOAA opened the pull request:
This PR, a companion to GDASApp PR [#983](https://github.com/NOAA-EMC/GDASApp/pull/983), creates a new Rocoto job called "atmanlfv3inc" that computes the FV3 atmosphere increment from the JEDI variational increment using a JEDI OOPS app in GDASApp, called fv3jedi_fv3inc.x, that replaces the GDASApp Python script, jediinc2fv3.py, for the variational analysis. The "atmanlrun" job is renamed "atmanlvar" to better reflect the role it plays of running now one of two JEDI executables for the atmospheric analysis jobs.

# Description

Previously, the JEDI variational executable would interpolate and write its increment, during the atmanlrun job, to the Gaussian grid, and then the python script, jediinc2fv3.py, would read it and then write the FV3 increment on the Gaussian grid during the atmanlfinal job. Following the new changes, the JEDI increment will be written directly to the cubed sphere. Then during the atmanlfv3inc job, the OOPS app will read it and compute the FV3 increment directly on the cubed sphere and write it out onto the Gaussian grid.

The reason for writing first to the cubed sphere grid is that otherwise the OOPS app would have to interpolate twice, once from Gaussian to cubed sphere before computing the increment and then back to the Gaussian, since all the underlying computations in JEDI are done on the native grid.

The motivation for this new app and job is that eventually we wish to transition all intermediate data to the native cubed sphere grid, and the OOPS framework allows us the flexibility to read and write to/from any grid format we wish by just changing the YAML configuration file rather than hardcoding. When we do switch to the cubed sphere, it will be an easy transition. Moreover, it the computations the OOPS app will be done with a compiled executable rather than an interpreted Python script, providing some performance increase.

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES
- Does this change require a documentation update? NO

# How has this been tested?
It has been tested with a cycling experiment with JEDI in both Hera and Orion to show that it runs without issues, and I have compared the FV3 increments computed by the original and news codes. The delp and hydrostatic delz increments, the key increments produced during this step, differ by a relative error of 10^-7 and 10^-2 respectively. This difference is most likely due to the original python script doing its internal computation on the interpolated Gaussian grid, while the new OOPS app does its computations on the native cubed sphere before interpolating the the Gaussian grid.

# Checklist
- [ ] Any dependent changes have been merged and published
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

DavidNew-NOAA committed:
Adding rocoto job for converting JEDI increment to FV3 increment

DavidNew-NOAA committed:
Added HPC environments and bug fixes

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Remove jediinc2fv3.py from link_workflow.sh

DavidNew-NOAA committed:
Fixed small omission

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Rename atmanlrun and atmanlinc to atmanlvar and atmanlfv3 respectively

DavidNew-NOAA committed:
Merge branch 'feature/jediinc2fv3' of https://github.com/DavidNew-NOAA/global-workflow into feature/jediinc2fv3

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA commented:
Closing until changes made

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Finalized HPC environment file changes and bugs in analysis scripts

DavidNew-NOAA committed:
Minor documentation update

DavidNew-NOAA committed:
Undo change

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Fix trailing whitespace

DavidNew-NOAA committed:
Revert changes that completely eliminate jediinc2fv3.py interactions since still needed for ensemble analysis job

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Update GDASApp submodule with PR #983

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

aerorahul commented on parm/config/gfs/config.atmanl:
```diff
@@ -6,7 +6,8 @@
 
 export OBS_LIST="${PARMgfs}/gdas/atm/obs/lists/gdas_prototype_3d.yaml.j2"
-export JEDIYAML="${PARMgfs}/gdas/atm/variational/3dvar_drpcg.yaml.j2"
+export VARYAML="${PARMgfs}/gdas/atm/variational/3dvar_drpcg.yaml.j2"
```
Renaming  to `VARYAML` and the associated changes will most certainly break the snow and aerosol tests.

`JEDIYAML` and `JEDIEXE` are required for all analysis tasks.
The issue we now face is that we are calling multiple executables (and their associated yamls) while using the same `analysis` class.

RussTreadon-NOAA replied:
C96C48_ufs_hybatmDA CI tests on Orion are failing in enkfgdasatmensanlinit.  The traceback suggests the failure is related to this change and related changes in `analysis.py` and/or `atm_analysis.py`.  Not sure yet if the solution is to modify `atmens_analysis.py` and `config.atmensanl`, modify some changes in this PR, a combination of both, or something else.

aerorahul commented on parm/config/gfs/config.atmanl:
```diff
@@ -28,6 +29,7 @@ export layout_y_atmanl=@LAYOUT_Y_ATMANL@
 
-export JEDIEXE=${EXECgfs}/fv3jedi_var.x
+export VAREXE=${EXECgfs}/fv3jedi_var.x
+export FV3INCEXE=${EXECgfs}/fv3jedi_fv3inc.x
```
This could become `JEDIEXE` in `config.atmanlfv3inc`

DavidNew-NOAA replied:
@aerorahul For consistency then, why not define JEDIEXE=${EXECgfs}/fv3jedi_var.x in config.atmanlvar (formerly config.atmanlrun) rather than config.atmanl?

aerorahul commented on parm/config/gfs/config.atmanl:
```diff
@@ -6,7 +6,8 @@
 export OBS_LIST="${PARMgfs}/gdas/atm/obs/lists/gdas_prototype_3d.yaml.j2"
-export JEDIYAML="${PARMgfs}/gdas/atm/variational/3dvar_drpcg.yaml.j2"
+export VARYAML="${PARMgfs}/gdas/atm/variational/3dvar_drpcg.yaml.j2"
+export FV3INCYAML="${PARMgfs}/gdas/atm/utils/fv3jedi_fv3inc.yaml.j2"
```
This could become `JEDIYAML` in `config.atmanlfv3inc`

aerorahul commented on parm/config/gfs/config.atmanlfv3inc:
```diff
@@ -0,0 +1,11 @@
+
+# Get task specific resources
+. "${EXPDIR}/config.resources" atmanlfv3inc
+
```
How about define `JEDIYAML` and `JEDIEXE` in this script for the yaml and executable used by this job.

aerorahul commented on ush/python/pygfs/task/analysis.py:
```diff
@@ -43,12 +43,15 @@ def initialize(self) -> None:
-        # link jedi executable to run directory
-        self.link_jediexe()
+        # link jedi variational executable to run directory
+        self.link_varexe()
```
I think this should be reverted.

aerorahul commented on ush/python/pygfs/task/analysis.py:
```diff
@@ -43,12 +43,15 @@ def initialize(self) -> None:
+        self.link_varexe()
+
+        # link fv3 increment converter executable to run directory
+        self.link_fv3incexe()
```
This should be called may be in the `ex-script` or may be create a method say `initialize_fv3inc` and then call through there

DavidNew-NOAA commented on ush/python/pygfs/task/atm_analysis.py:
```diff
@@ -100,9 +103,15 @@ def initialize(self: Analysis) -> None:
+
+        # generate FV3 increment converter YAML file
+        logger.debug(f"Generate FV3 increment converter YAML file: {self.task_config.fv3inc_yaml}")
+        fv3inc_yaml = parse_j2yaml(self.task_config.FV3INCYAML, self.task_config, searchpath=self.gdasapp_j2tmpl_dir)
```
@aerorahul The problem with changing FV3INCYAML to JEDIYAML in config.atmanlfv3 is that, in the init job here, JEDIYAML is already the variational yaml. It's possible instead that I could stage the FV3 increment converter yaml in the the atmanlfv3inc job itself rather than the atmanlinit job.

DavidNew-NOAA replied:
The problem with that though is I will be using a job dedicated to running an executable to stage a YAML, and I thought the purpose of atmanlinit and atmanlfinal was for things like moving around files.

DavidNew-NOAA committed:
Reverting accidental renaming

RussTreadon-NOAA commented:
**Orion tests using C96C48_ufs_hybatmDA CI**

`/work2/noaa/da/rtreadon/git/global-workflow/test` contains 
 - working copy of `feature/jediinc2fv3` at c23c7439
 - `sorc/gdas.cd` is GDASApp `develop` at [`3c9cdaa`](https://github.com/NOAA-EMC/GDASApp/commit/3c9cdaaca7836c8fa1787696a6ba4a5dbb736082)

Both g-w and `gdas.cd` contain local modifications which enable g-w C96C48_ufs_hybatmDA CI to successfully run JEDI variational and local ensemble DA jobs.  The EXPDIR for JEDI ATM CI is `/work2/noaa/stmp/rtreadon/EXPDIR/prtest`

Local modifications in the working copy of `feature/jediinc2fv3` are
- `ci/cases/pr/C96C48_ufs_hybatmDA.yaml` - activate this CI case on Orion
- `parm/config/gfs/config.atmanl` - replace `VARYAML` with `JEDIYAML`, replace `VAREXE` with `JEDIEXE`, remove `FV3INCEXE`
- `parm/config/gfs/config.atmanlfv3inc` - add `FV3INCYAML` and `FV3INCEXE`
- `scripts/exglobal_atm_analysis_variational.py` - replace `AtmAnl.variational()` with `AtmAnl.execute()`
- `scripts/exgdas_enkf_sfc.sh` - change in g-w PR #2443 required for enkfgdasesfc to successfully run to completion
- `ush/python/pygfs/task/analysis.py` - revert this back to `develop` version of this script but retain `link_fv3incexe`
- `ush/python/pygfs/task/atm_analysis.py` - replace `var_yaml` with `jedi_yaml`
- `workflow/applications/applications.py` - replace `atmanlrun` with `atmanlvar` and `atmanlfv3inc`

I am not entirely satisfied with the above changes.  There are some inconsistencies and sloppiness which need to be cleaned up.  This is just a first effort to get things working with refinements to be made in subsequent iterations.

Local modifications in `gdas.cd` are
- `parm/atm/obs/lists/gdas_prototype_3d.yaml.j2` - remove seviri observations (see GDASApp issue [#1008](https://github.com/NOAA-EMC/GDASApp/issues/1008))
- `parm/atm/utils/fv3jedi_fv3inc.yaml.j2` - use `_anl` dimensions for increment

The gdas and gfs analcalc jobs fail.  Executable `calc_analysis.x` aborts when processing ensemble resolution files
```
 Copying from Background phalf
 Adding Increment to clwmr  liq_wat
 rank of data array != variable ndims (or ndims-1)
 rank of data array != variable ndims (or ndims-1)
```
The new oops increment code creates fields with a time dimension: `(time, lev, lat, lon)`.  The python script created increment fields without a time dimension: `(lev, lat, lon)`.  Operational `atmi003.nc`, `atminc.nc`, and `atmi009.nc` do not include the time dimension.  Operational global increment fields are written as `(lev, lat, lon)`.  

I added another dimension to the increment work array in calc_analysis source code file `inc2anl.f90`, recompiled, and reran analcalc.  This allowed the failed ensemble resolution execution of `calc_analysis.x` to run to completion.  The modified `inc2anl.f90` is in `/work2/noaa/da/rtreadon/git/global-workflow/test/sorc/gsi_utils.fd/src/netcdf_io/calc_analysis.fd`

Feel free to poke around in `/work2/noaa/da/rtreadon/git/global-workflow/test` and use what you want.

DavidNew-NOAA committed:
Reverting some changes and revising some things.

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA commented:
@RussTreadon-NOAA Thanks for the very useful feedback. My latest commit is mostly similar changes, except I kept AtmAnl.variational() rather than AtmAnl.execute(), because there's now two execution methods in the atm analysis class, one for variational and one for increment conversion. 

I also put the linking of the increment converter executable directly into the initialization method for the atm analysis class as well. Otherwise, If I retained it in the init method for the general analysis class then that executable would get linked for other analysis jobs like aerosols. I didn't fully understand the analysis job design philosophy until now.

I'm running out of time until the weekend, but I will add your changes the the analcalc job next week. Note: the next gdasfcst job ran fine with the increment having a time dimensions. I'm not sure why it didn't fail if it's the case that normally there is not time dimension in the FV3 increment.

Enjoy your weekend and time off.

DavidNew-NOAA committed:
Update GDASApp develop

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Update name of FV3 increment converter YAML for variational job

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Slight refactoring of analysis jobs

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Move executable linking to executation functions in analysis jobs

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Coding norms

DavidNew-NOAA committed:
Revert some things

DavidNew-NOAA committed:
Hopefully final changes and reversions

DavidNew-NOAA committed:
Fix calcanl_gfs.py to allow time dimension in increment. Also, update gsi_utils and gdasapp submodules

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Coding norms

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA commented:
I've made change that address @aerorahul concerns related to the design philosophy of the gdas atmospheric analysis jobs and how to implement this new job. I also made a companion [PR](https://github.com/NOAA-EMC/GSI-utils/pull/35) for gsi_utils that allows one to specify in the calc_analysis.x namelist whether the input increment is in GSI or JEDI format, that is, whether or not it has a time dimension. If "jedi" is set to .true. in this namelist (default is false), the change that @RussTreadon-NOAA came up with, which adds another dimension to the increment working array, is applied. 

This namelist parameter is now set in ush/calc_analysis.py in G-W, and controlled by whether DO_JEDIATMVAR is set to "YES" or "NO" in config.base.

RussTreadon-NOAA commented on env/HERCULES.env:
```diff
@@ -53,13 +53,13 @@ case ${step} in
     export wave_mpmd=${mpmd_opt}
 
  ;;
- "atmanlrun")
```
CI test C96C48_ufs_hybatmDA failed on Hercules because `HERCULES.env` does not contain a section for `atmanlfv3inc`.  

@AndrewEichmann-NOAA , `ocnanalecen` is also missing from `HERCULES.env`.

RussTreadon-NOAA reviewed (commented):
GDASApp `develop` was updated this morning (4/15) to [c2fdf1e](https://github.com/NOAA-EMC/GDASApp/commit/c2fdf1e652b8bb367e1e33780acc8f94811a25f5).  Should we update `gdas.cd` to point at this GDASApp hash?

GSI-utils PR [#35](https://github.com/NOAA-EMC/GSI-utils/pull/35) was merged into GSI-utils `develop` at [68bc14d](https://github.com/NOAA-EMC/GSI-utils/commit/68bc14d30b3ca8f890f2761c8bdd0a3cea635cf1).  We should update the `gsi_utils.fd` submodule to point at this hash.

DavidNew-NOAA committed:
Update GDASApp hash

DavidNew-NOAA committed:
Add atmanlfv3inc to HERCULES.env

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA commented:
@RussTreadon-NOAA I added atmanlfv3inc to HERCULES.env and updated the GDASApp hash to most recent develop with the fixes to the jjob testing and increment converter. The gsi_utils submodule is already up-to-date.

DavidNew-NOAA committed:
Update GSI Utils hash

DavidNew-NOAA commented:
Actually looks like I had the wrong GSI Utils hash. Updated now

DavidNew-NOAA committed:
Make placeholder in AWSPW.env and CONTAINER.env for atmanlfv3inc job

RussTreadon-NOAA reviewed (approved):
Thank you @DavidNew-NOAA for the updates.   The following has been done
1. Recompile on Hera
2. Run test_gdasapp.  All tests pass except test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY.  This is a known failure (see GDASApp issue [#994](https://github.com/NOAA-EMC/GDASApp/issues/994))
3. Rerun C96C48_hybatmDA (GSI-based CI).  All jobs run to completion.
4. Rerun C96C48_ufs_hybatmDA (JEDI-based CI).  All jobs run to completion given the local modifications below
   - JEDI-based DA requires `qoption` change to `fix/gdas/gsibec/C48/gfs_gsi_global.nml`
   - JEDI-based DA requires removal of `seviri` from GDASApp `parm/atm/obs/lists/gdas_prototype_3d.yaml.j2`.  GDASApp PR [#1043](https://github.com/NOAA-EMC/GDASApp/pull/1043) merged this change into GDASApp `develop` at [fd1ebda](https://github.com/NOAA-EMC/GDASApp/commit/fd1ebdad33245b7d9184a1972e1c73d5caf3881d).  You could update `gdas.cd` in this PR to [fd1ebda](https://github.com/NOAA-EMC/GDASApp/commit/fd1ebdad33245b7d9184a1972e1c73d5caf3881d).  Not a requirement, just a consideration.

Approve.

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

CoryMartin-NOAA commented:
@DavidNew-NOAA is this ready for final review? I see @RussTreadon-NOAA 's tests look good

DavidNew-NOAA commented:
@CoryMartin-NOAA Yes, I think it's ready.

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA committed:
Merge branch 'feature/jediinc2fv3' of https://github.com/DavidNew-NOAA/global-workflow into feature/jediinc2fv3

CoryMartin-NOAA commented:
@DavidNew-NOAA can you update the GDAS hash to develop and push a new commit?

DavidNew-NOAA committed:
Update GDASApp hash

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

DavidNew-NOAA commented:
@CoryMartin-NOAA Hash updated

aerorahul commented on parm/config/gfs/config.atmanlvar:
```diff
@@ -0,0 +1,11 @@
+
+# Get task specific resources
+. "${EXPDIR}/config.resources" atmanlvar
+
```
Did we not discuss moving the JEDI_YAML and JEDI_EXE for the variational executable here from `config.atmanl`?

DavidNew-NOAA replied:
@aerorahul We did, but I reverted a bunch of changes that weren't directly related to the atmanlfv3inc job, because this PR was becoming pretty unwieldy. I can change that line back though if you want.

DavidNew-NOAA replied:
@aerorahul Additionally though, recall from our discussion that only the JEDIEXE can be moved, not the JEDIYAML, because both the init and var run job need the variational YAML.

aerorahul replied:
lets keep this as is for now, but make an issue to discuss and hash this out sooner (rather than later).

DavidNew-NOAA replied:
Will do

aerorahul reviewed (approved):
Looks good.
We are going to run this through CI.

WalterKolczynski-NOAA commented:
Error log: [gist](https://gist.github.com/emcbot/9ebb22dc251def113a266f486566453a)

RussTreadon-NOAA commented:
The failed job, [gdasprepsnowobs.log](https://gist.github.com/emcbot/9ebb22dc251def113a266f486566453a#file-gdasprepsnowobs-log), encountered during CI-Hera is not altered by this PR.   This PR impacts JEDI ATM DA (specifically JEDI ATM VAR), not JEDI SNOW DA.   

@DavidNew-NOAA & @CoryMartin-NOAA - please correct me if I am wrong.

CoryMartin-NOAA commented:
yes @RussTreadon-NOAA I have reverted the GDASApp PR that caused this issue, we need the tests ran again with the latest hash of GDASApp develop

DavidNew-NOAA commented:
@RussTreadon-NOAA Correct, snow DA is not touched by this PR.

RussTreadon-NOAA commented:
Thank you both for your quick replies.  Hopefully we can move this PR forward and get it into `develop` soon.

aerorahul committed:
Merge branch 'develop' into feature/jediinc2fv3

aerorahul commented:
@DavidNew-NOAA Can you please update the pointer to gdasapp from this AM's commit after [fc62ef5](https://github.com/NOAA-EMC/GDASApp/commit/fc62ef5f4ddcd2c13df515b593e5ce1c7b5dd10b) and we can kick off the test again.

DavidNew-NOAA committed:
Update GDASApp hash

DavidNew-NOAA commented:
@aerorahul hash updated

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/jediinc2fv3

RussTreadon-NOAA commented:
**Orion tests**
Complete the following on Orion:

1. Install `DavidNew-NOAA:feature/jediinc2fv3` at [94d28d56](https://github.com/DavidNew-NOAA/global-workflow/commit/94d28d566b3db272d4e1ec2dbe4cb40f810b6182) in `HOMEgfs=/work2/noaa/da/rtreadon/git/global-workflow/jediinc2fv3`
2. `cd $HOMEgfs/sorc`
3. `./build_all.sh -g -u -v`
4. `link_workflow.sh`

Enable and run g-w JEDI ATM CI as follows:
1. `cd $HOMEgfs/ci/cases/pr`.  Remove `- orion` entry from `C96C48_ufs_hybatmDA.yaml`
2. `cd /work2/noaa/stmp/rtreadon`
3. edit `setup_ci.sh` to point at above `$HOMEgfs`.  Execute `./setup_ci.sh`.  Create `EXPDIR` and `COMROOT` for `prtest` in which to run C96C48 JEDI ATM cycle.
4. activate crontab to `rocotorun prtest`

The 2024022318 half cycle is complete.  The 2024022400 gdas and enkfgdas cycle is complete.  JEDI ATM jobs ran successfully in the gdas, enkfgdas, and gfs cycles
```
202402240000      gdasprepatmiodaobs                    17549303           SUCCEEDED                   0         1         362.0
202402240000          gdasatmanlinit                    17549333           SUCCEEDED                   0         1          66.0
202402240000           gdasatmanlvar                    17549356           SUCCEEDED                   0         1         614.0
202402240000        gdasatmanlfv3inc                    17549385           SUCCEEDED                   0         1          68.0
202402240000         gdasatmanlfinal                    17549387           SUCCEEDED                   0         1         141.0
...
202402240000    enkfgdasatmensanlinit                    17549334           SUCCEEDED                   0         1          66.0
202402240000    enkfgdasatmensanlrun                    17549357           SUCCEEDED                   0         1         478.0
202402240000    enkfgdasatmensanlfinal                    17549379           SUCCEEDED                   0         1         130.0
...
202402240000       gfsprepatmiodaobs                    17549304           SUCCEEDED                   0         1         367.0
202402240000           gfsatmanlinit                    17549335           SUCCEEDED                   0         1          70.0
202402240000            gfsatmanlvar                    17549358           SUCCEEDED                   0         1         516.0
202402240000         gfsatmanlfv3inc                    17549380           SUCCEEDED                   0         1          75.0
202402240000          gfsatmanlfinal                    17549386           SUCCEEDED                   0         1         111.0
```

Enable and run GDASApp `test_gdasapp` as follows:
1. `cd $HOMEgfs/sorc/gdas.cd`
2. `module use $HOMEgfs/sorc/gdas.cd/modulefiles`
3. `module load GDAS/orion.intel`
4. set environment variables for `test_gdasapp`
6. `cd $HOMEgfs/sorc/gdas.cd/build`
7. `ctest -R test_gdasapp`

All 45 tests pass.
```
Test project /work/noaa/da/rtreadon/git/global-workflow/jediinc2fv3/sorc/gdas.cd/build
      Start 1393: test_gdasapp_util_coding_norms
 1/54 Test #1393: test_gdasapp_util_coding_norms ........................   Passed    1.92 sec
      Start 1394: test_gdasapp_util_ioda_example
 2/54 Test #1394: test_gdasapp_util_ioda_example ........................   Passed    3.24 sec
      Start 1395: test_gdasapp_util_prepdata
 3/54 Test #1395: test_gdasapp_util_prepdata ............................   Passed    1.00 sec
      Start 1396: test_gdasapp_util_rads2ioda
 4/54 Test #1396: test_gdasapp_util_rads2ioda ...........................   Passed    0.09 sec
      Start 1397: test_gdasapp_util_ghrsst2ioda
 5/54 Test #1397: test_gdasapp_util_ghrsst2ioda .........................   Passed    0.06 sec
      Start 1398: test_gdasapp_util_smap2ioda
 6/54 Test #1398: test_gdasapp_util_smap2ioda ...........................   Passed    0.06 sec
      Start 1399: test_gdasapp_util_smos2ioda
 7/54 Test #1399: test_gdasapp_util_smos2ioda ...........................   Passed    0.06 sec
      Start 1400: test_gdasapp_util_viirsaod2ioda
 8/54 Test #1400: test_gdasapp_util_viirsaod2ioda .......................   Passed    0.06 sec
      Start 1401: test_gdasapp_util_icecamsr2ioda
 9/54 Test #1401: test_gdasapp_util_icecamsr2ioda .......................   Passed    0.06 sec
      Start 1739: test_gdasapp_check_python_norms
10/54 Test #1739: test_gdasapp_check_python_norms .......................   Passed    2.68 sec
      Start 1740: test_gdasapp_check_yaml_keys
11/54 Test #1740: test_gdasapp_check_yaml_keys ..........................   Passed    0.06 sec
      Start 1741: test_gdasapp_jedi_increment_to_fv3
12/54 Test #1741: test_gdasapp_jedi_increment_to_fv3 ....................   Passed    0.32 sec
      Start 1742: test_gdasapp_setup_cycled_exp
13/54 Test #1742: test_gdasapp_setup_cycled_exp .........................   Passed    4.98 sec
      Start 1743: test_gdasapp_fv3jedi_fv3inc
14/54 Test #1743: test_gdasapp_fv3jedi_fv3inc ...........................   Passed   22.19 sec
      Start 1744: test_gdasapp_convert_bufr_temp_dbuoy
15/54 Test #1744: test_gdasapp_convert_bufr_temp_dbuoy ..................   Passed    0.16 sec
      Start 1745: test_gdasapp_convert_bufr_salt_dbuoy
16/54 Test #1745: test_gdasapp_convert_bufr_salt_dbuoy ..................   Passed    0.13 sec
      Start 1746: test_gdasapp_convert_bufr_temp_mbuoyb
17/54 Test #1746: test_gdasapp_convert_bufr_temp_mbuoyb .................   Passed    0.15 sec
      Start 1747: test_gdasapp_convert_bufr_salt_mbuoyb
18/54 Test #1747: test_gdasapp_convert_bufr_salt_mbuoyb .................   Passed    0.13 sec
      Start 1748: test_gdasapp_convert_bufr_tesacprof
19/54 Test #1748: test_gdasapp_convert_bufr_tesacprof ...................   Passed    0.14 sec
      Start 1749: test_gdasapp_convert_bufr_trkobprof
20/54 Test #1749: test_gdasapp_convert_bufr_trkobprof ...................   Passed    0.13 sec
      Start 1750: test_gdasapp_convert_bufr_sfcships
21/54 Test #1750: test_gdasapp_convert_bufr_sfcships ....................   Passed    0.13 sec
      Start 1751: test_gdasapp_convert_bufr_sfcshipsu
22/54 Test #1751: test_gdasapp_convert_bufr_sfcshipsu ...................   Passed    0.13 sec
      Start 1752: test_gdasapp_soca_nsst_increment_to_mom6
23/54 Test #1752: test_gdasapp_soca_nsst_increment_to_mom6 ..............   Passed   62.23 sec
      Start 1753: test_gdasapp_soca_prep
24/54 Test #1753: test_gdasapp_soca_prep ................................   Passed    1.60 sec
      Start 1754: test_gdasapp_soca_run_clean
25/54 Test #1754: test_gdasapp_soca_run_clean ...........................   Passed    0.01 sec
      Start 1755: test_gdasapp_soca_setup_obsprep
26/54 Test #1755: test_gdasapp_soca_setup_obsprep .......................   Passed   29.45 sec
      Start 1756: test_gdasapp_soca_JGLOBAL_PREP_OCEAN_OBS
27/54 Test #1756: test_gdasapp_soca_JGLOBAL_PREP_OCEAN_OBS ..............   Passed   52.85 sec
      Start 1757: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
28/54 Test #1757: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP ....   Passed   74.14 sec
      Start 1758: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_BMAT
29/54 Test #1758: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_BMAT ....   Passed   42.14 sec
      Start 1759: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN
30/54 Test #1759: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN .....   Passed   42.14 sec
      Start 1760: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN
31/54 Test #1760: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN ....   Passed   42.16 sec
      Start 1761: test_gdasapp_soca_copy_scratch
32/54 Test #1761: test_gdasapp_soca_copy_scratch ........................   Passed    0.43 sec
      Start 1762: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT
33/54 Test #1762: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT ...   Passed   42.14 sec
      Start 1763: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_POST
34/54 Test #1763: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_POST ....   Passed   10.14 sec
      Start 1764: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY
35/54 Test #1764: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY ....   Passed  170.14 sec
      Start 1765: test_gdasapp_soca_socahybridweights
36/54 Test #1765: test_gdasapp_soca_socahybridweights ...................   Passed   42.12 sec
      Start 1766: test_gdasapp_soca_incr_handler
37/54 Test #1766: test_gdasapp_soca_incr_handler ........................   Passed    2.10 sec
      Start 1767: test_gdasapp_soca_ens_handler
38/54 Test #1767: test_gdasapp_soca_ens_handler .........................   Passed   10.10 sec
      Start 1768: test_gdasapp_snow_create_ens
39/54 Test #1768: test_gdasapp_snow_create_ens ..........................   Passed    0.92 sec
      Start 1769: test_gdasapp_snow_imsproc
40/54 Test #1769: test_gdasapp_snow_imsproc .............................   Passed    2.01 sec
      Start 1770: test_gdasapp_snow_apply_jediincr
41/54 Test #1770: test_gdasapp_snow_apply_jediincr ......................   Passed    2.60 sec
      Start 1771: test_gdasapp_snow_letkfoi_snowda
42/54 Test #1771: test_gdasapp_snow_letkfoi_snowda ......................   Passed    8.37 sec
      Start 1772: test_gdasapp_convert_bufr_adpsfc_snow
43/54 Test #1772: test_gdasapp_convert_bufr_adpsfc_snow .................   Passed    2.28 sec
      Start 1773: test_gdasapp_convert_bufr_adpsfc
44/54 Test #1773: test_gdasapp_convert_bufr_adpsfc ......................   Passed    2.99 sec
      Start 1774: test_gdasapp_convert_gsi_satbias
45/54 Test #1774: test_gdasapp_convert_gsi_satbias ......................   Passed    1.35 sec
      Start 1775: test_gdasapp_setup_atm_cycled_exp
46/54 Test #1775: test_gdasapp_setup_atm_cycled_exp .....................   Passed    0.85 sec
      Start 1776: test_gdasapp_atm_jjob_var_init
47/54 Test #1776: test_gdasapp_atm_jjob_var_init ........................   Passed   44.11 sec
      Start 1777: test_gdasapp_atm_jjob_var_run
48/54 Test #1777: test_gdasapp_atm_jjob_var_run .........................   Passed  106.16 sec
      Start 1778: test_gdasapp_atm_jjob_var_inc
49/54 Test #1778: test_gdasapp_atm_jjob_var_inc .........................   Passed   42.15 sec
      Start 1779: test_gdasapp_atm_jjob_var_final
50/54 Test #1779: test_gdasapp_atm_jjob_var_final .......................   Passed   42.16 sec
      Start 1780: test_gdasapp_atm_jjob_ens_init
51/54 Test #1780: test_gdasapp_atm_jjob_ens_init ........................   Passed   43.90 sec
      Start 1781: test_gdasapp_atm_jjob_ens_run
52/54 Test #1781: test_gdasapp_atm_jjob_ens_run .........................   Passed  266.18 sec
      Start 1782: test_gdasapp_atm_jjob_ens_final
53/54 Test #1782: test_gdasapp_atm_jjob_ens_final .......................   Passed   42.17 sec
      Start 1783: test_gdasapp_aero_gen_3dvar_yaml
54/54 Test #1783: test_gdasapp_aero_gen_3dvar_yaml ......................   Passed    0.72 sec

100% tests passed, 0 tests failed out of 54

Label Time Summary:
gdas-utils    =   6.55 sec*proc (9 tests)
script        =   6.55 sec*proc (9 tests)

Total Test time (real) = 1271.07 sec
```

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA , what else remains to be done before this PR can be merged into `develop`?

TerrenceMcGuinness-NOAA commented:
![image](https://github.com/NOAA-EMC/global-workflow/assets/66784684/3be0c0d2-c7ad-430f-9bdc-5c081252560f)
@RussTreadon-NOAA The Continuous Integration tests to check the PR agains the develop build and run base line almost passed, but we ran against discrepancy in our validation of success that produced a false **_negative_** in this last case.  We are working to fix that now.

RussTreadon-NOAA commented:
Thank you @TerrenceMcGuinness-NOAA for sharing what's going on.  Does this problem prevent merger of this PR into `develop`?

TerrenceMcGuinness-NOAA commented:
Yeah, @RussTreadon-NOAA that's the role of the Branch Manger of course, but not sure why it would be a problem.

aerorahul committed:
Merge branch 'develop' into feature/jediinc2fv3

aerorahul committed:
Merge branch 'develop' into feature/jediinc2fv3
